core/src/main/scala/com.softwaremill.kmq/redelivery/RedeliverActor.scala

```diff
@@ -1,45 +1,55 @@
 package com.softwaremill.kmq.redelivery
 
-import akka.actor.Actor
-import com.softwaremill.kmq.MarkerKey
-import com.typesafe.scalalogging.StrictLogging
+import cats.effect.std.Queue
+import cats.effect.{IO, Resource}
+import com.softwaremill.kmq.{KafkaClients, KmqConfig, MarkerKey}
 
 import scala.concurrent.duration._
-
-class RedeliverActor(p: Partition, redeliverer: Redeliverer) extends Actor with StrictLogging {
-
-  private var toRedeliver: List[MarkerKey] = Nil
-
-  import context.dispatcher
-
-  override def preStart(): Unit = {
-    logger.info(s"${self.path} Started redeliver actor for partition $p")
+import fs2.Stream
+import org.apache.kafka.clients.producer.KafkaProducer
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+object RedeliverActor {
+
+  private val logger = Slf4jLogger.getLogger[IO]
+  def create(
+      p: Partition,
+      producer: KafkaProducer[Array[Byte], Array[Byte]],
+      config: KmqConfig,
+      clients: KafkaClients
+  ): Resource[IO, RedeliverActor] = {
+    val resource = for {
+      redeliverer <- Resource.make(
+        IO(new RetryingRedeliverer(new DefaultRedeliverer(p, producer, config, clients)))
+      )(r => IO(r.close()))
+      mailbox <- Resource.eval(Queue.unbounded[IO, RedeliverActorMessage])
+      _ <- Resource.eval(logger.info(s"Started redeliver actor for partition $p"))
+      _ <- stream(redeliverer, mailbox).compile.drain.background
+    } yield new RedeliverActor(mailbox)
+
+    resource.onFinalize(
+      logger.info(s"Stopped redeliver actor for partition $p")
+    )
   }
 
-  override def postStop(): Unit = {
-    try redeliverer.close()
-    catch {
-      case e: Exception => logger.error(s"Cannot close redeliverer for partition $p", e)
+  private def stream(redeliverer: Redeliverer, mailbox: Queue[IO, RedeliverActorMessage]) =
+    Stream.fromQueueUnterminated(mailbox).evalMapAccumulate(List.empty[MarkerKey]) {
+      case (toRedeliver, RedeliverMarkers(m)) =>
+        IO.pure((toRedeliver ++ m, ()))
+      case (toRedeliver, DoRedeliver) =>
+        val hadRedeliveries = toRedeliver.nonEmpty
+        IO(redeliverer.redeliver(toRedeliver))
+          .as((Nil, ()))
+          .guarantee {
+            if (hadRedeliveries)
+              mailbox.offer(DoRedeliver)
+            else
+              mailbox.offer(DoRedeliver).delayBy(1.second)
```

Comment:
Isn't there a `.start` missing at the end? otherwise the `guarantee` will take up to a second, waiting for this effect to complete - we won't handle other `RedelvierMakrers` messages meanwhile

Reply:
correct, fixed it